runtime/integration-tests/src/generic/envs/runtime_env.rs

```diff
@@ -0,0 +1,200 @@
+use std::{cell::RefCell, marker::PhantomData, rc::Rc};
+
+use cfg_primitives::{Address, BlockNumber, Header, Index};
+use codec::Encode;
+use cumulus_primitives_core::PersistedValidationData;
+use cumulus_primitives_parachain_inherent::ParachainInherentData;
+use cumulus_test_relay_sproof_builder::RelayStateSproofBuilder;
+use frame_support::{
+	inherent::{InherentData, ProvideInherent},
+	storage::transactional,
+};
+use sp_consensus_aura::{Slot, AURA_ENGINE_ID};
+use sp_core::H256;
+use sp_runtime::{
+	generic::{Era, SignedPayload},
+	traits::{Block, Extrinsic},
+	Digest, DigestItem, DispatchError, DispatchResult, MultiSignature, Storage, TransactionOutcome,
+};
+use sp_timestamp::Timestamp;
+
+use crate::{
+	generic::{
+		environment::{Blocks, Env},
+		runtime::Runtime,
+	},
+	utils::accounts::Keyring,
+};
+
+pub struct RuntimeEnv<T: Runtime> {
+	nonce: Index,
+	ext: Rc<RefCell<sp_io::TestExternalities>>,
+	_config: PhantomData<T>,
+}
+
+impl<T: Runtime> Env<T> for RuntimeEnv<T> {
+	fn from_storage(storage: Storage) -> Self {
+		let mut ext = sp_io::TestExternalities::new(storage);
+
+		ext.execute_with(|| Self::prepare_block(1));
+
+		Self {
+			nonce: 0,
+			ext: Rc::new(RefCell::new(ext)),
+			_config: PhantomData,
+		}
+	}
+
+	fn submit(&mut self, who: Keyring, call: impl Into<T::RuntimeCall>) -> DispatchResult {
+		self.ext.borrow_mut().execute_with(|| {
+			let runtime_call = call.into();
+			let signed_extra = (
+				frame_system::CheckNonZeroSender::<T>::new(),
+				frame_system::CheckSpecVersion::<T>::new(),
+				frame_system::CheckTxVersion::<T>::new(),
+				frame_system::CheckGenesis::<T>::new(),
+				frame_system::CheckEra::<T>::from(Era::mortal(256, 0)),
+				frame_system::CheckNonce::<T>::from(self.nonce),
+				frame_system::CheckWeight::<T>::new(),
+				pallet_transaction_payment::ChargeTransactionPayment::<T>::from(0),
+			);
+
+			let raw_payload =
+				SignedPayload::new(runtime_call.clone(), signed_extra.clone()).unwrap();
+			let signature =
+				MultiSignature::Sr25519(raw_payload.using_encoded(|payload| who.sign(payload)));
+
+			let multi_address = (Address::Id(who.to_account_id()), signature, signed_extra);
+
+			let extrinsic =
+				<T::Block as Block>::Extrinsic::new(runtime_call, Some(multi_address)).unwrap();
+
+			self.nonce += 1;
+
+			T::apply_extrinsic(extrinsic).unwrap()
+		})
+	}
+
+	fn pass(&mut self, blocks: Blocks<T>) {
+		self.ext.borrow_mut().execute_with(|| {
+			let next = frame_system::Pallet::<T>::block_number() + 1;
+
+			let last_block = match blocks {
```

Comment:
This name can be a bit confusing since it's not getting the last block but the block at which we want to stop at. Within context it makes sense but I had to think that twice. Just a thought 👍

Reply:
Maybe `block_to_stop` is more readable?

Reply:
yeah or `end_block`. something like that 👍

Reply:
I like it more!